### PR TITLE
Charger detection

### PR DESCRIPTION
**Pull request by nwdepatie**
## Changes

Basically created a latching state that uses interrupts to set whether the BMS is on car or on charger cart. Also tried to copy over some more stuff from the charging branch

## Notes

This is a lil jank but like eh. I bypassed the CAN queue thing to only detect the presence of CAN traffic

## Test Cases

- Not tested yet :) Will test on charging cart now that we have better flashing

### CODE REVIEW

File: Core/Src/compute.c [hunk 1]
```diff
@@ -121,20 +127,26 @@ int compute_send_charging_message(uint16_t voltage_to_set, uint16_t current_to_s
...
 	}
-	#endif
 
 	return 0;
 }
 
+void compute_set_charger_connected()
+{
+	if (!is_on_car)
+		is_charger_connected = true;;
```
Comment (dyldonahue) [inline Core/Src/compute.c:143]: Nit but extra ;

File: Core/Src/compute.c [hunk 2]
```diff
@@ -98,12 +102,14 @@ int compute_send_charging_message(uint16_t voltage_to_set, uint16_t current_to_s
 	} charger_msg_data;
 
 	charger_msg_data.charger_voltage = voltage_to_set * 10;
-	charger_msg_data.charger_current = current_to_set * 10;
+	// Override charging to charge at 3.5A (or 0 if charging disabled)
+	charger_msg_data.charger_current = current_to_set == 0 ? 0 : 35;
+	printf("Current Requested: %d \r\n", charger_msg_data.charger_current);
```
Comment (dyldonahue) [inline Core/Src/compute.c:107]: Note that with good charger this should be 5 amps, we lowered to 3.5 for Burlington. Not sure what capabilities are at comp

**Review by dyldonahue — APPROVED** [review]
Looks good, hopefully can callback behavior plays nicer here

File: Core/Src/main.c [hunk 3]
```diff
@@ -16,7 +16,7 @@
 
 #include "segment.h"
 #include "compute.h"
-#include "datastructs.h"
+#include "datastructs.h"CAN_RX_FIFO1
```
Comment (jr1221) [inline Core/Src/main.c]: @Sabramz what does this line mean?

File: Core/Src/stm32f4xx_it.c [hunk 4]
```diff
@@ -220,28 +240,17 @@ void DMA2_Stream0_IRQHandler(void)
 void CAN2_RX0_IRQHandler(void)
 {
   /* USER CODE BEGIN CAN2_RX0_IRQn 0 */
-  can_receive_callback(&hcan2);
+  //can_receive_callback(&hcan2);
+
+  /* If ANY CAN msg is received on CAN 2, we can assume it is on charger cart */
+  compute_set_charger_connected();
```
Comment (jr1221) [inline Core/Src/stm32f4xx_it.c:246]: Note the first can message sent is **after** the charger is enabled so we have no way of knowing charge current until after the car starts. I can change this via TPU if needed by making charger box TPU send a message. Kinda scary imo.

File: shepherd2.ioc [hunk 5]
```diff
@@ -143,11 +143,11 @@ Mcu.PinsNb=55
 Mcu.ThirdPartyNb=0
 Mcu.UserConstants=
 Mcu.UserName=STM32F405RGTx
-MxCube.Version=6.10.0
-MxDb.Version=DB.6.0.100
+MxCube.Version=6.11.0
```
Comment (jr1221) [inline shepherd2.ioc]: Please revert this it breaks most dev computers.

File: Core/Src/main.c (hunk 3 above)
Comment (dyldonahue) [reply Core/Src/main.c]: this looks like a bug

Comment (jr1221) [reply Core/Src/main.c]: How did it still compile?

File: Core/Src/stm32f4xx_it.c (hunk 4 above)
Comment (dyldonahue) [reply Core/Src/stm32f4xx_it.c:246]: not sure i understand your point here

File: Core/Src/main.c (hunk 3 above)
Comment (dylan-donahue) [reply Core/Src/main.c:19]: guessing just gave a warning

Comment (dyldonahue) [reply Core/Src/main.c]: probably just gave a warning

**Pull request closed without merging**